### PR TITLE
Patcher Fixes

### PR DESCRIPTION
**Pull request by Expl0dingBanana**
If the current patch level was removed, the patcher was starting at the first patch level versus the next one
There is now a Primary Key check on the versions table.  If it does not exist, it will attempt to add it.  If there are duplicate keys in the table, it will de-dupe then apply the primary key
trs_status will perform more logic to the upgrade to reduce the chances of losing the information in that table

### CODE REVIEW

File: mapadroid/patcher/__init__.py [hunk 1]
```diff
@@ -37,9 +36,23 @@ def __init__(self, args, data_manager):
...
-        if self.__update_required():
-            self.__update_mad()
+        self._installed_ver = None
+        self.__validate_versions_schema()
+        # If the versions table does not exist we have to install the schema.  When the schema is set, we are then on
+        # the latest table so none of these checks are required
+        if self._installed_ver is None:
+            self.__convert_mappings()
+            self.__get_installed_version()
+            # TODO - When should these be executed?  Can the be added somewhere and ignored?
```
Comment (fosJoddie) [inline mapadroid/patcher/__init__.py]: I'm not comfortable adding in a TODO in a PR - the position they are in looks ok as long as none of the patches ever expect these tables and columns to be in a different state than "newest" - if they do, they shouldn't be called "unversioned" (obviously)

Comment (Expl0dingBanana) [reply mapadroid/patcher/__init__.py]: That to-do needed to be dropped. During initial development I dropped that in to remind myself to validate where to put them. I'll remove that to-do line

**Comment by krzee** [comment]
before this patch i got:
`
  File "<string>", line 2, in autofetch_column
  File "/usr/lib/python3.6/multiprocessing/managers.py", line 772, in _callmethod
    raise convert_to_error(kind, result)
          │                │     └ ProgrammingError("1054 (42S22): Unknown column 'instance' in 'where clause'", "1054 (42S22): Unknown column 'instance' in 'wh...
          │                └ '#ERROR'
          └ <function convert_to_error at 0x7fd0be8707b8>

mysql.connector.errors.ProgrammingError: 1054 (42S22): Unknown column 'instance' in 'where clause'
`
after this patch i got:
`
  File "/root/MAD/mapadroid/patcher/__init__.py", line 216, in __update_mad
    self.__apply_update(patch_ver)
    │                   └ 23
    └ <mapadroid.patcher.MADPatcher object at 0x7f6b3d00d828>

> File "/root/MAD/mapadroid/patcher/__init__.py", line 66, in __apply_update
    patch = patch_base.Patch(logger, self.dbwrapper, self.data_manager, self._application_args)
            │          │     │       │    │          │    │             │    └ Namespace(DEBUG=False, adb_server_ip='127.0.0.1', adb_server_port=5037, advanced_config=True, auto_hatch=False, auto_hatch_nu...
            │          │     │       │    │          │    │             └ <mapadroid.patcher.MADPatcher object at 0x7f6b3d00d828>
            │          │     │       │    │          │    └ <mapadroid.utils.data_manager.DataManager object at 0x7f6b3d00d5c0>
            │          │     │       │    │          └ <mapadroid.patcher.MADPatcher object at 0x7f6b3d00d828>
            │          │     │       │    └ <mapadroid.db.DbWrapper.DbWrapper object at 0x7f6b3d05da90>
            │          │     │       └ <mapadroid.patcher.MADPatcher object at 0x7f6b3d00d828>
            │          │     └ <loguru.logger handlers=[(id=1, level=20, sink=<stdout>), (id=2, level=40, sink=<stderr>), (id=3, level=20, sink='logs/202001...
            │          └ <class 'mapadroid.patcher.patch_23.Patch'>
            └ <module 'mapadroid.patcher.patch_23' from '/root/MAD/mapadroid/patcher/patch_23.py'>

  File "/root/MAD/mapadroid/patcher/_patch_base.py", line 36, in __init__
    self._execute()
    │    └ <function Patch._execute at 0x7f6b3cf95510>
    └ <mapadroid.patcher.patch_23.Patch object at 0x7f6b3d0475f8>

  File "/root/MAD/mapadroid/patcher/patch_23.py", line 89, in _execute
    if not device_id:

NameError: name 'device_id' is not defined
`

**Comment by krzee** [comment]
I had another server to update. removing trs_status did not help, but this patch did. I got the same error as above but then I ran it again and it continued.

**Comment by Expl0dingBanana** [comment]
Rebased to latest master.  Updates for last 2 patches:
- Fix the issue identified by krzee
- Fix an issue that I have seen in chat where the trs_status table might not have the correct columns due to a failed upgrade that succeeded.  If v23 or v24 and the columns dont exist, roll back to 22 and re-apply.  The trs_status table will be preserved.

**Pull request merged**